Training/interface/DataLoader_main.h

```diff
@@ -0,0 +1,840 @@
+      end_entry = std::min((long long)end_dataset, tauTuple->GetEntries());
+
+      // histogram to calculate weights
+      // TH1::AddDirectory(kFALSE);
+      TFile* file_input = new TFile(input_spectrum.c_str());
+      TFile* file_target = new TFile(target_spectrum.c_str());
+      TH2D* target_hist = (TH2D*)file_target->Get("eta_pt_hist_tau");
+
+      for(int type = 0; type < 4; ++type) {
+        hist_weights.push_back(shared_ptr<TH2D>(new TH2D(("w_1_"+tau_types_names[type]).c_str(),
+                                                         ("w_1_"+tau_types_names[type]).c_str(),
+                                                         n_eta_bins, eta_min, eta_max,
+                                                         n_pt_bins, pt_min, pt_max)));
+        hist_weights[type]->SetDirectory(0); // disabling the file referencing
+
+        TH2D* after_rebin_input_hist = new TH2D("input_hist", "input_hist",
+                                                n_eta_bins, eta_min, eta_max,
+                                                n_pt_bins, pt_min, pt_max);
+        TH2D* input_hist = (TH2D*)file_input->Get(("eta_pt_hist_"+tau_types_names[type]).c_str());
+
+        RebinAndFill(*hist_weights[type], *target_hist);
+        RebinAndFill(*after_rebin_input_hist, *input_hist);
+        hist_weights[type]->Divide(after_rebin_input_hist);
+        delete after_rebin_input_hist;
+      }
+      file_input->Close();
+      file_target->Close();
+      MaxDisbCheck(hist_weights, weight_thr);
+    }
+
+    bool MoveNext() {
+        if(!tauTuple)
+          throw std::runtime_error("DataLoader is not initialized.");
+
+        data = std::make_shared<Data>(n_tau, n_fe_tau, n_inner_cells, n_outer_cells,
+                                      n_pf_el, n_pf_mu, n_pf_chHad, n_pf_nHad,
+                                      n_pf_gamma, n_ele, n_muon, tau_types
+                                      );
+
+        const Long64_t end_point = current_entry + n_tau;
+        size_t n_processed = 0, n_total = static_cast<size_t>(end_point - current_entry);
+        for(Long64_t tau_i = 0; tau_i < n_tau; ++current_entry) {
+          if(current_entry == end_entry) return false;
+          tauTuple->GetEntry(current_entry);
+          const auto& tau = tauTuple->data();
+          // skip event if it is not tau_e, tau_mu, tau_jet or tau_h
+          if(tau.tauType < 0 || tau.tauType > 3) continue;
+          else {
+            data->y_onehot[ tau_i * tau_types + tau.tauType ] = 1.0; // filling labels
+            data->weight.at(tau_i) = GetWeight(tau.tauType, tau.tau_pt, std::abs(tau.tau_eta)); // filling weights
+            FillTauBranches(tau, tau_i, data);
+            FillCellGrid(tau, tau_i, innerCellGridRef, data, true);
+            FillCellGrid(tau, tau_i, outerCellGridRef, data, false);
+            ++tau_i;
+          }
+        }
+        hasData = true;
+        return true;
+    }
+
```

Comment:
We also need the method `Reset` that sets current_entry to 0. It will be used to run training for more than one epoch.